File: android-native/app/src/main/java/com/example/kpkn/data/learn/LearnData.kt
package com.example.kpkn.data.learn

import androidx.compose.ui.graphics.Color
import kotlinx.serialization.Serializable

// ─── Categorías de Learn ─────────────────────────────────────────────────────

enum class LearnCategory(
    val label: String,
    val color: Color,
    val icon: String,
) {
    BEGINNER("Novato", Color(0xFF00BCD4), "rocket"),
    TRAINING("Entrenamiento", Color(0xFF1E88E5), "dumbbell"),
    NUTRITION("Nutrición", Color(0xFF43A047), "apple"),
    RINGS("RINGS", Color(0xFF448AFF), "ring"),
    MENTAL_HEALTH("Salud Mental", Color(0xFF7E57C2), "mind"),
    TOOLS("Herramientas", Color(0xFFFF8F00), "wrench"),
}

// ─── Tipos de contenido ──────────────────────────────────────────────────────

enum class ContentType {
    PARAGRAPH,
    HEADING,
    BULLET,
    TIP,
    WARNING,
    CALLOUT,
}

data class ContentBlock(
    val type: ContentType,
    val text: String = "",
    val items: List<String> = emptyList(),
    val accentColor: Long? = null, // ARGB para CALLOUT
)

// ─── Quiz ────────────────────────────────────────────────────────────────────

data class QuizQuestion(
    val id: String,
    val question: String,
    val options: List<String>,
    val correctIndex: Int,
    val explanation: String,
)

// ─── Curso ───────────────────────────────────────────────────────────────────

data class LearnModule(
    val id: String,
    val title: String,
    val category: LearnCategory,
    val shortDescription: String,
    val icon: String,
    val estimatedMinutes: Int,
    val disclaimer: String? = null,
    val submodules: List<LearnSubmodule>,
    val finalQuiz: List<QuizQuestion>,
    val isSpecial: Boolean = false, // true solo para "Tu primera rutina en KPKN"
)

data class LearnSubmodule(
    val id: String,
    val title: String,
    val content: List<ContentBlock>,
    val quiz: List<QuizQuestion>,
)

// ─── Insignias ───────────────────────────────────────────────────────────────

@Serializable
enum class BadgeTier { BRONCE, PLATA, ORO }

@Serializable
data class LearnBadge(
    val courseId: String,
    val courseName: String,
    val tier: BadgeTier?,
    val earnedAt: String,
    val score: Int,
    val isSpecial: Boolean = false, // "Graduado KPKN"
)

// ─── Progreso ────────────────────────────────────────────────────────────────

@Serializable
data class CourseProgress(
    val moduleId: String,
    val submoduleIndex: Int = -1,               // último submódulo completado
    val submoduleQuizScores: Map<String, Int> = emptyMap(), // submoduleId -> score/10
    val finalQuizScore: Int = -1,               // -1 = no rendido
    val badge: LearnBadge? = null,
    val completedAt: String? = null,
    val disclaimerShown: Boolean = false,
) {
    val isCompleted: Boolean get() = finalQuizScore >= 8
    val inProgress: Boolean get() = submoduleIndex >= 0 || submoduleQuizScores.isNotEmpty()
    val totalSubmodulesCompleted: Int get() = submoduleQuizScores.size
}

File: ios-native/KPKNFit/KPKNFit/Data/Learn/LearnData.swift
import Foundation

public struct LearnCategory: Codable, Equatable {
    public let rawValue: String
    public let label: String
    public let color: String
    public let icon: String

    public init(rawValue: String, label: String, color: String, icon: String) {
        self.rawValue = rawValue
        self.label = label
        self.color = color
        self.icon = icon
    }

    public static let BEGINNER = LearnCategory(rawValue: "BEGINNER", label: "Novato", color: "0xFF00BCD4", icon: "rocket")
    public static let TRAINING = LearnCategory(rawValue: "TRAINING", label: "Entrenamiento", color: "0xFF1E88E5", icon: "dumbbell")
    public static let NUTRITION = LearnCategory(rawValue: "NUTRITION", label: "Nutrición", color: "0xFF43A047", icon: "apple")
    public static let RINGS = LearnCategory(rawValue: "RINGS", label: "RINGS", color: "0xFF448AFF", icon: "ring")
    public static let MENTAL_HEALTH = LearnCategory(rawValue: "MENTAL_HEALTH", label: "Salud Mental", color: "0xFF7E57C2", icon: "mind")
    public static let TOOLS = LearnCategory(rawValue: "TOOLS", label: "Herramientas", color: "0xFFFF8F00", icon: "wrench")
}

public enum ContentType: String, Codable {
    case PARAGRAPH
    case HEADING
    case BULLET
    case TIP
    case WARNING
    case CALLOUT
}

public struct ContentBlock: Codable {
    public let type: ContentType
    public let text: String
    public let items: [String]
    public let accentColor: String?

    public init(type: ContentType, text: String = "", items: [String] = [], accentColor: String? = nil) {
        self.type = type
        self.text = text
        self.items = items
        self.accentColor = accentColor
    }
}

public struct QuizQuestion: Codable {
    public let id: String
    public let question: String
    public let options: [String]
    public let correctIndex: Int
    public let explanation: String

    public init(id: String, question: String, options: [String], correctIndex: Int, explanation: String) {
        self.id = id
        self.question = question
        self.options = options
        self.correctIndex = correctIndex
        self.explanation = explanation
    }
}

public struct LearnSubmodule: Codable {
    public let id: String
    public let title: String
    public let content: [ContentBlock]
    public let quiz: [QuizQuestion]

    public init(id: String, title: String, content: [ContentBlock], quiz: [QuizQuestion]) {
        self.id = id
        self.title = title
        self.content = content
        self.quiz = quiz
    }
}

public enum BadgeTier: String, Codable {
    case BRONCE
    case PLATA
    case ORO
}

public struct LearnBadge: Codable {
    public let courseId: String
    public let courseName: String
    public let tier: BadgeTier?
    public let earnedAt: String
    public let score: Int
    public let isSpecial: Bool

    public init(courseId: String, courseName: String, tier: BadgeTier?, earnedAt: String, score: Int, isSpecial: Bool = false) {
        self.courseId = courseId
        self.courseName = courseName
        self.tier = tier
        self.earnedAt = earnedAt
        self.score = score
        self.isSpecial = isSpecial
    }
}

public struct CourseProgress: Codable {
    public let moduleId: String
    public var submoduleIndex: Int
    public var submoduleQuizScores: [String: Int]
    public var finalQuizScore: Int
    public var badge: LearnBadge?
    public var completedAt: String?
    public var disclaimerShown: Bool

    public init(moduleId: String, submoduleIndex: Int = -1, submoduleQuizScores: [String: Int] = [:], finalQuizScore: Int = -1, badge: LearnBadge? = nil, completedAt: String? = nil, disclaimerShown: Bool = false) {
        self.moduleId = moduleId
        self.submoduleIndex = submoduleIndex
        self.submoduleQuizScores = submoduleQuizScores
        self.finalQuizScore = finalQuizScore
        self.badge = badge
        self.completedAt = completedAt
        self.disclaimerShown = disclaimerShown
    }

    public var isCompleted: Bool { finalQuizScore >= 8 }
    public var inProgress: Bool { submoduleIndex >= 0 || !submoduleQuizScores.isEmpty }
    public var totalSubmodulesCompleted: Int { submoduleQuizScores.count }

    public func copy(
        moduleId: String? = nil,
        submoduleIndex: Int? = nil,
        submoduleQuizScores: [String: Int]? = nil,
        finalQuizScore: Int? = nil,
        badge: LearnBadge? = nil,
        completedAt: String? = nil,
        disclaimerShown: Bool? = nil
    ) -> CourseProgress {
        CourseProgress(
            moduleId: moduleId ?? self.moduleId,
            submoduleIndex: submoduleIndex ?? self.submoduleIndex,
            submoduleQuizScores: submoduleQuizScores ?? self.submoduleQuizScores,
            finalQuizScore: finalQuizScore ?? self.finalQuizScore,
            badge: badge ?? self.badge,
            completedAt: completedAt ?? self.completedAt,
            disclaimerShown: disclaimerShown ?? self.disclaimerShown
        )
    }
}

public struct LearnModule: Codable {
    public let id: String
    public let title: String
    public let category: LearnCategory
    public let shortDescription: String
    public let icon: String
    public let estimatedMinutes: Int
    public let disclaimer: String?
    public let submodules: [LearnSubmodule]
    public let finalQuiz: [QuizQuestion]
    public let isSpecial: Bool

    public init(id: String, title: String, category: LearnCategory, shortDescription: String, icon: String, estimatedMinutes: Int, disclaimer: String? = nil, submodules: [LearnSubmodule], finalQuiz: [QuizQuestion] = [], isSpecial: Bool = false) {
        self.id = id
        self.title = title
        self.category = category
        self.shortDescription = shortDescription
        self.icon = icon
        self.estimatedMinutes = estimatedMinutes
        self.disclaimer = disclaimer
        self.submodules = submodules
        self.finalQuiz = finalQuiz
        self.isSpecial = isSpecial
    }
}
